src/common/type/pack.c

```diff
@@ -0,0 +1,1571 @@
+        FUNCTION_TEST_PARAM(BOOL, param.defaultNull);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    if (!pckWriteDefaultNull(this, param.defaultNull, value == NULL))
+        pckWriteTag(this, pckTypePtr, param.id, (uintptr_t)value);
+
+    FUNCTION_TEST_RETURN(this);
+}
+
+/**********************************************************************************************************************************/
+PackWrite *
+pckWriteStr(PackWrite *this, const String *value, PckWriteStrParam param)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_WRITE, this);
+        FUNCTION_TEST_PARAM(STRING, value);
+        FUNCTION_TEST_PARAM(UINT, param.id);
+        FUNCTION_TEST_PARAM(BOOL, param.defaultNull);
+        FUNCTION_TEST_PARAM(STRING, param.defaultValue);
+    FUNCTION_TEST_END();
+
+    if (!pckWriteDefaultNull(this, param.defaultNull, strEq(value, param.defaultValue)))
+    {
+        ASSERT(value != NULL);
+
+        pckWriteTag(this, pckTypeStr, param.id, strSize(value) > 0);
+
+        if (strSize(value) > 0)
+        {
+            pckWriteUInt64Internal(this, strSize(value));
+            pckWriteBuffer(this, BUF(strZ(value), strSize(value)));
+        }
+    }
+
+    FUNCTION_TEST_RETURN(this);
+}
+
+/**********************************************************************************************************************************/
+PackWrite *
+pckWriteTime(PackWrite *this, time_t value, PckWriteTimeParam param)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_WRITE, this);
+        FUNCTION_TEST_PARAM(TIME, value);
+        FUNCTION_TEST_PARAM(UINT, param.id);
+        FUNCTION_TEST_PARAM(BOOL, param.defaultNull);
+        FUNCTION_TEST_PARAM(TIME, param.defaultValue);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    if (!pckWriteDefaultNull(this, param.defaultNull, value == param.defaultValue))
+        pckWriteTag(this, pckTypeTime, param.id, cvtInt64ToZigZag(value));
+
+    FUNCTION_TEST_RETURN(this);
+}
+
+/**********************************************************************************************************************************/
+PackWrite *
+pckWriteU32(PackWrite *this, uint32_t value, PckWriteUInt32Param param)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(PACK_WRITE, this);
+        FUNCTION_TEST_PARAM(UINT32, value);
+        FUNCTION_TEST_PARAM(UINT, param.id);
+        FUNCTION_TEST_PARAM(BOOL, param.defaultNull);
+        FUNCTION_TEST_PARAM(UINT32, param.defaultValue);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+
+    if (!pckWriteDefaultNull(this, param.defaultNull, value == param.defaultValue))
+        pckWriteTag(this, pckTypeU32, param.id, value);
+
+    FUNCTION_TEST_RETURN(this);
+}
+
+/**********************************************************************************************************************************/
+PackWrite *
+pckWriteU64(PackWrite *this, uint64_t value, PckWriteUInt64Param param)
```

Comment:
There no way to know sure sure that defaultValue is set because it could be set to zero.

But, we could automatically set defaultNull when defaultValue is non 0/NULL/false. So you could set either defaultNull or defaultValue and it would work. Would that be better or would it make things even more confusing?

Reply:
Or we have a separate functions for defaults, e.g. `pckReadU64DefaultP()`.

Reply:
defaultNull = true means 0/NULL/false, right? So if I say defaultNull = true, why would I need defaultValue? I think it would be much clearer to only set detaultValue if we do not want to store the default when it is something other than that that is classified as defaultNull. I personally don't like defaultValue, but if we're going to use it, I think it would be much clearer to have each of these parameters have a clear and separate purpose (even if we have to set defaultNull behind the scenes).

Reply:
> Sure, I'm OK with updating defaultNull behind the scenes.

After a brief discussion 12/7/2020, defaultNull may become defaultWrite and treating it as writing the default into the pack when set to true.

Reply:
Done in 522ae2eb3dfe.